File: src/cycod/Helpers/DirectoryHelpers.cs
using System;
using System.Collections.Generic;
using System.IO;
using System.Linq;
using System.Text;
using System.Text.RegularExpressions;

public class DirectoryHelpers
{
    public static string FindOrCreateDirectorySearchParents(params string[] paths)
    {
        return FindDirectorySearchParents(paths, createIfNotFound: true)!;
    }

    public static string? FindDirectorySearchParents(params string[] paths)
    {
        return FindDirectorySearchParents(paths, createIfNotFound: false);
    }

    public static string? FindDirectorySearchParents(string[] paths, bool createIfNotFound)
    {
        var current = Directory.GetCurrentDirectory();
        while (current != null)
        {
            var combined = Path.Combine(paths.Prepend(current).ToArray());
            if (Directory.Exists(combined))
            {
                return combined;
            }

            current = Directory.GetParent(current)?.FullName;
        }

        if (createIfNotFound)
        {
            current = Directory.GetCurrentDirectory();
            var combined = Path.Combine(paths.Prepend(current).ToArray());
            return EnsureDirectoryExists(combined);
        }

        return null;
    }
    
    public static string EnsureDirectoryExists(string folder)
    {
        try
        {
            var validFolderName = !string.IsNullOrEmpty(folder);
            if (validFolderName && !Directory.Exists(folder))
            {
                Directory.CreateDirectory(folder!);
            }
        }
        catch (Exception ex)
        {
            if (!Directory.Exists(folder))
            {
                Console.WriteLine($"Error creating directory: {ex.Message}");
                throw;
            }
        }
        return folder;
    }

    public static void EnsureDirectoryForFileExists(string fileName)
    {
        EnsureDirectoryExists(Path.GetDirectoryName(fileName) ?? ".");
    }
}

File: src/commonjs/Helpers/DirectoryHelpers.ts
import * as fs from 'fs';
import * as path from 'path';
import { ConsoleHelpers } from './ConsoleHelpers';

/**
 * Directory operations helper class
 */
export class DirectoryHelpers {
    public static findOrCreateDirectorySearchParents(...paths: string[]): string
    {
        return DirectoryHelpers.findDirectorySearchParentsWithCreate(paths, true)!;
    }

    public static findDirectorySearchParents(...paths: string[]): string | null
    {
        return DirectoryHelpers.findDirectorySearchParentsWithCreate(paths, false);
    }

    public static findDirectorySearchParentsWithCreate(paths: string[], createIfNotFound: boolean): string | null
    {
        let current = process.cwd();
        
        while (current != null) {
            const combined = path.join(current, ...paths);
            if (fs.existsSync(combined) && fs.statSync(combined).isDirectory()) {
                return combined;
            }

            const parent = path.dirname(current);
            current = parent !== current ? parent : null;
        }

        if (createIfNotFound) {
            current = process.cwd();
            const combined = path.join(current, ...paths);
            return DirectoryHelpers.ensureDirectoryExists(combined);
        }

        return null;
    }
    
    public static ensureDirectoryExists(folder: string): string
    {
        try {
            const validFolderName = folder != null && folder !== '';
            if (validFolderName && !fs.existsSync(folder)) {
                fs.mkdirSync(folder, { recursive: true });
            }
        } catch (ex) {
            if (!fs.existsSync(folder)) {
                ConsoleHelpers.writeErrorLine(`Error creating directory: ${(ex as Error).message}`);
                throw ex;
            }
        }
        return folder;
    }

    public static ensureDirectoryForFileExists(fileName: string): void
    {
        DirectoryHelpers.ensureDirectoryExists(path.dirname(fileName) || '.');
    }
}
